openstack-auto-scaling.sh

```diff
@@ -0,0 +1,349 @@
+  log "0" "Expected number of hypervisors in aggregation group $agg_group up: $computes_number_up_after_logic"
+  log "0" "Hypervisors in aggregation group $agg_group up: $(echo $computes_hostnames_up)"
+  log "0" "Hypervisors in aggregation group $agg_group down: $(echo $computes_hostnames_down_with_no_vm)"
+}
+
+computes_shutdown() {
+#function expects global variables
+  local computes_hostnames_to_disable
+  computes_hostnames_to_disable="$(echo "$computes_hostnames_up_available_for_shutdown"|tail -"$computes_number_difference")"
+  local computes_hostnames_to_shutdown
+  computes_hostnames_to_shutdown="$(echo "$computes_hostnames_up_available_for_shutdown"|tail -"$computes_number_difference"| \
+                                            awk -F\. '{print $1}')"
+  print_agg_group_status
+  if [ "$TESTMODE" == "false" ]; then
+    log "0" "Disabling nova-compute service for hypervisors: $(echo $computes_hostnames_to_disable) now.."
+    echo "$computes_hostnames_to_disable"| while read -r host; do 
+                                             openstack_overcloud compute service set --disable --disable-reason "$SCRIPT_NAME" "$host" nova-compute
+                                           done
+    log "0" "Shutting down hypervisors: $(echo $computes_hostnames_to_disable) now.."
+    echo "$computes_hostnames_to_shutdown"| while read -r host; do
+                                              openstack_undercloud server stop "$host"
+                                            done
+    log "0" "Waiting for hypervisors to be down with timeout $SHUTDOWN_TIMEOUT seconds"
+    SECONDS="clear";
+    local filter
+    filter="$(echo "$computes_hostnames_to_disable"|tr '\n' '|'|sed 's/|$//g')"
+    while [ "$SECONDS" -lt "$SHUTDOWN_TIMEOUT" ]; do
+      openstack_overcloud hypervisor list --long -f value -c "Hypervisor Hostname" -c "State"|grep -E "$filter"|grep " up$" &>/dev/null || break
+      sleep 10
+    done
+    openstack_overcloud hypervisor list --long -f value -c "Hypervisor Hostname" -c "State"|grep -E "$filter"|grep " up$" &>/dev/null
+    local RETVAL
+    RETVAL="$?"
+    if [ "$RETVAL" -eq 0 ]; then
+      compute_hostnames_to_disable_still_up="$(openstack_overcloud hypervisor list --long -f value -c "Hypervisor Hostname" -c "State"| \
+                                                 grep -E "$filter"|grep " up$"|awk '{print $1}'|sort)"
+      log "1" "Computes still up after timeout: $( echo $compute_hostnames_to_disable_still_up)"
+    fi
+    log "0" "Disabling nova-compute service for hypervisors: $(echo $computes_hostnames_to_disable) now.."
+    echo "$computes_hostnames_to_disable"| while read -r host; do openstack_overcloud compute service set --enable "$host" nova-compute; done
+  else
+    log "2" "Script in TEST MODE. No action is taken on hypervisors"
+    log "0" "Hypervisors for shutdown: $(echo $computes_hostnames_to_disable)"
+  fi
+}
+
+computes_startup() {
```

Comment:
this should be gone to  a single function to do both tasks to avoid copy/pasting code